StereographicProductManifold to use gyrovector space functions in product manifolds

gatoniel opened the pull request:
Hi,

I subclassed the `ProductManifold` and created a `StereographicProductManifold`. Here I added the functions `dist2plane`, `expmap0 `and `mobius_add` by calling the respective functions in the underlying `Stereographic` manifolds. I added some test functions for this.
Additionally, I added `wrapped_normal` to `Stereographic` as an alternative random function and I added scipy to the requirements in setup.py as mentioned in #161

rrkarim commented:
@gatoniel could we also add the test for toy optimization task on `StereographicProductManifold` as it is done for other manifolds. Just to be sure the numerical stability is good.

gatoniel commented:
@rrkarim yes, I can do that. Could you tell me which test or files exactly I should change? Is it in https://github.com/geoopt/geoopt/blob/master/tests/test_manifold_basic.py ? And should I also add tests in test_origin.py, test_random.py, test_adam.py ?

rrkarim commented:
There are some simple optimization tasks on different manifolds in `test_adam.py`. You could add the test for `StereographicProductManifold` there or for another optimization method.

gatoniel commented:
Can I do something else to put forward this PR?

ferrine commented:
I'll review the pr today. My concern is that it is something we already support by the rest of geoopt functionality

ferrine commented:
Just to be on the same page. The motivation to add this specific manifold extension is to support math functions on the product manifold? I remember suffering with that so this motivation is super valid.

gatoniel commented:
@ferrine yes, that´s the only motivation. For example, I want to calculate the distance between points on a PoincareBall x StereographicProjection. I think it is easier to do this with one simple call to the ProductManifold than writing my own wrapper function that extracts the submanifolds and calculates the distances per manifold...

I just implemented the functions that I needed immediately. But it might be good to also add all other gyro vector functions. I can do this when I am less busy with other stuff.

ferrine commented on geoopt/manifolds/product.py:
```diff
@@ -435,3 +437,96 @@ def origin(
+        self,
+        *manifolds_with_shape: Tuple[Stereographic, Union[Tuple[int, ...], int]],
+    ):
+        super().__init__(*manifolds_with_shape)
```
I'd add a check after the parent init to verify all manifolds are the same family

ferrine replied:
```suggestion
        super().__init__(*manifolds_with_shape)
        for man in self.manifilds:
            if not geoopt.utils.ismanifold(man, Stereographic):
                raise TypeError("Every submanifold as to be Stereographic manifold")
```

ferrine commented on geoopt/manifolds/product.py:
```diff
@@ -435,3 +437,96 @@ def origin(
+            mapped_tensors.append(mapped)
+        return self.pack_point(*mapped_tensors)
+
+    def dist2plane(
```
Need to figure out scaling requirements here. Should be similar to how scaling is declared in Stereographic manifold

ferrine replied:
In short scaling is used to rescale tangent space coordinates

ferrine commented on geoopt/manifolds/product.py:
```diff
@@ -435,3 +437,96 @@ def origin(
+        self,
+        *manifolds_with_shape: Tuple[Stereographic, Union[Tuple[int, ...], int]],
+    ):
+        super().__init__(*manifolds_with_shape)
```
I'd add a check to make sure all manifolds are of the same family

ferrine commented on geoopt/manifolds/stereographic/manifold.py:
```diff
@@ -418,6 +449,53 @@ def random_normal(
 
     random = random_normal
 
+    def wrapped_normal(
```
Did not find a coverage test for that

ferrine replied:
We also need a scaling rule for std
```suggestion
@__scaling__(ScalingInfo(std=-1), "random")
    def wrapped_normal(
```

ferrine commented on geoopt/manifolds/stereographic/manifold.py:
```diff
@@ -418,6 +449,53 @@ def random_normal(
 
     random = random_normal
 
+    def wrapped_normal(
```
Did not find a coverage test for that

ferrine reviewed (changes requested):
Some minor stuff is left, but it is critical before merge, otherwise looks good.

ferrine commented on geoopt/manifolds/product.py:
```diff
@@ -435,3 +437,96 @@ def origin(
+    >>> import geoopt
+    >>> sphere = geoopt.SphereProjection()
+    >>> torus = StereographicProductManifold((sphere, 2), (sphere, 2))
+    """
```
This will apply the same scaling rules for new functions as they were in the Stereographic manifold. Basically, all you'll need
```suggestion
    """
    __scaling__ = Stereographic.__scaling__.copy()
```

ferrine commented on geoopt/manifolds/stereographic/manifold.py:
```diff
@@ -418,6 +449,53 @@ def random_normal(
     random = random_normal
 
+    def wrapped_normal(
+        self, mean: torch.Tensor, std, size, dtype=None, device=None
```
Make API similar to `normal`
```suggestion
        self, *size, mean: torch.Tensor, std=1, dtype=None, device=None
```

ferrine reviewed (changes requested):
I've left suggestions for the code to meet the dev API requirements, you can apply and test them

gatoniel commented:
Hi, thanks. I´ll try to get to it next week.

ferrine commented:
great

On Thu, 15 Apr 2021 at 10:25, niklas netter ***@***.***>
wrote:

> Hi, thanks. I´ll try to get to it next week.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/geoopt/geoopt/pull/163#issuecomment-820185434>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACZJX3V7EUIHEMLLRWD77RDTI2IHFANCNFSM4Z6ADO7A>
> .
>


-- 
Best, Max

My timezone: GMT+3

gatoniel commented:
My newest commit is pushed to my repo but it does not appear here...

gatoniel commented:
I just encountered this error when using the `StereographicProductManifold` with `Scaled`:

> AttributeError: 'StereographicProductManifold' object has no attribute 'geodesic_unit'

I will add another commit for this.

ferrine commented:
Hey, do you need any help with the PR?

gatoniel commented:
Hi, I am sorry. I was busy with other stuff. I will try to fix the mentioned error above and create a new commit in the next 3 weeks.

ferrine commented:
Nice

On Sun, 18 Jul 2021, 12:43 niklas netter, ***@***.***> wrote:

> Hi, I am sorry. I was busy with other stuff. I will try to fix the
> mentioned error above and create a new commit in the next 3 weeks.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/geoopt/geoopt/pull/163#issuecomment-882029953>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACZJX3RC7G5I3PPDY4SIQNTTYKO5XANCNFSM4Z6ADO7A>
> .
>

ferrine commented:
Hi, @gatoniel, how are you doing? Do you have time to continue with the PR?

gatoniel commented:
@ferrine sorry for the long delay. I added now the `geodesic_unit`. I will additionally add all the other functions available for `Stereographic` manifolds.

gatoniel commented:
@ferrine I added now all functions that are available for `Stereographic` manifolds to the `StereoprahicProductManifold`. I added tests for each of them. Although, these tests are only the most obvious tests I came up with: they do not test anything sophisticated. I did not know how to test `mobius_fn_apply` or `mobius_fn_apply_chain`.

ferrine commented on geoopt/manifolds/product.py:
```diff
@@ -435,3 +437,283 @@ def origin(
+            if not geoopt.utils.ismanifold(man, Stereographic):
+                raise TypeError("Every submanifold has to be Stereographic manifold")
+
+    @__scaling__(ScalingInfo(1))
```
```suggestion
```

ferrine commented on geoopt/manifolds/product.py:
```diff
@@ -435,3 +437,283 @@ def origin(
+            res_list.append(res)
+        return self.pack_point(*res_list)
+
+    @__scaling__(ScalingInfo(t=-1))
```
```suggestion
```

ferrine commented on geoopt/manifolds/product.py:
```diff
@@ -435,3 +437,283 @@ def origin(
+            res_list.append(res)
+        return self.pack_point(*res_list)
+
+    @__scaling__(ScalingInfo(1))
```
```suggestion
```

ferrine commented on geoopt/manifolds/product.py:
```diff
@@ -435,3 +437,283 @@ def origin(
+            res = torch.unsqueeze(res, -1)
+        return res
+
+    @__scaling__(ScalingInfo(u=-1))
```
```suggestion
```

ferrine commented on geoopt/manifolds/product.py:
```diff
@@ -435,3 +437,283 @@ def origin(
+            )
+        return self.pack_point(*res)
+
+    # this does not yet work with scaling
+    @__scaling__(ScalingInfo.NotCompatible)
```
```suggestion
```

ferrine reviewed (changes requested):
I've left suggestions to got rid of all scaling instances in the product manifold since this [line](https://github.com/geoopt/geoopt/pull/163/files#diff-20d9c3a9e1ff5354ad5bf4569ab98557d87edb9697afce3a018aecf18201684dR455)

ferrine commented:
Done, great work!

gatoniel commented:
👍 Sorry again, for the delay
